File: swift/Sources/PreceiptsKit/Comments.swift
// Review comments: local drafts (notes-to-agent, never posted) and the
// clipboard formats that feed them to a coding agent. Port of the Rust
// core's comments.rs.
//
// Local comments live under `.git/preceipts/comments.json`, keyed by
// branch — review state, not repo content; they survive the app closing
// but not intentional cleanup. GitHub feedback renders through the same
// CommentContext so every comment shares the Copy to Clipboard affordance.

import Foundation

public enum CommentSide: String, Codable, Sendable {
    case old
    case new
}

public struct LocalComment: Codable, Sendable, Identifiable {
    public let id: UInt64
    public let path: String
    /// 1-based line number on `side` of the diff.
    public let line: Int
    public let side: CommentSide
    /// The diff line's text when the comment was written — quoted in the
    /// clipboard format and used to flag the comment as outdated later.
    public let lineText: String
    public var body: String
    /// Unix seconds; display formatting is the UI's job.
    public let createdAt: UInt64

    enum CodingKeys: String, CodingKey {
        case id, path, line, side, body
        case lineText = "line_text"
        case createdAt = "created_at"
    }
}

/// Everything needed to render one comment to the clipboard, regardless
/// of origin (local draft or GitHub).
public struct CommentContext {
    public let path: String
    public let line: Int
    public let lineText: String
    public let body: String
    /// Nil for local drafts; "octocat" / "coderabbit[bot]" for GitHub.
    public let author: String?

    public init(path: String, line: Int, lineText: String, body: String, author: String?) {
        self.path = path
        self.line = line
        self.lineText = lineText
        self.body = body
        self.author = author
    }
}

/// One comment as a markdown block with file:line context:
///
///     apps/next/components/foo.tsx:123
///     > const x = useMemo(...)
///     This memo is unnecessary — props are primitives.
///
public func formatComment(_ comment: CommentContext) -> String {
    var out = "\(comment.path):\(comment.line)\n"
    let quoted = comment.lineText.trimmingTrailingWhitespace()
    if !quoted.trimmingCharacters(in: .whitespaces).isEmpty {
        out += "> \(quoted)\n"
    }
    if let author = comment.author {
        out += "\u{2014} \(author): "
    }
    out += comment.body.trimmingCharacters(in: .whitespacesAndNewlines)
    out += "\n"
    return out
}

/// Many comments as one digest, grouped by file — a paste-ready worklist
/// for a coding agent.
public func formatDigest(_ comments: [CommentContext]) -> String {
    var byFile: [String: [String]] = [:]
    for comment in comments {
        byFile[comment.path, default: []].append(formatComment(comment))
    }
    var out = ""
    for path in byFile.keys.sorted() {
        out += "## \(path)\n\n"
        for block in byFile[path]! {
            out += block
            out += "\n"
        }
    }
    return out.trimmingTrailingWhitespace() + "\n"
}

// ------------------------------------------------------------------
// Store

public final class CommentStore {
    private struct StoreFile: Codable {
        /// branch name → comments. Drafts are review state for a branch's PR.
        var branches: [String: [LocalComment]] = [:]
        var nextId: UInt64 = 0

        enum CodingKeys: String, CodingKey {
            case branches
            case nextId = "next_id"
        }
    }

    private let file: URL
    private let branch: String
    private var data: StoreFile

    /// Open (or create) the store for a repo's git dir and current branch.
    public init(gitDir: URL, branch: String) throws {
        let dir = gitDir.appendingPathComponent("preceipts")
        try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        self.file = dir.appendingPathComponent("comments.json")
        self.branch = branch
        if let bytes = try? Data(contentsOf: file),
            let decoded = try? JSONDecoder().decode(StoreFile.self, from: bytes)
        {
            self.data = decoded
        } else {
            self.data = StoreFile()
        }
    }

    public var comments: [LocalComment] {
        data.branches[branch] ?? []
    }

    @discardableResult
    public func add(
        path: String, line: Int, side: CommentSide, lineText: String, body: String
    ) throws -> LocalComment {
        data.nextId += 1
        let comment = LocalComment(
            id: data.nextId,
            path: path,
            line: line,
            side: side,
            lineText: lineText,
            body: body,
            createdAt: UInt64(Date().timeIntervalSince1970))
        data.branches[branch, default: []].append(comment)
        try save()
        return comment
    }

    public func remove(id: UInt64) throws {
        data.branches[branch]?.removeAll { $0.id == id }
        try save()
    }

    public func updateBody(id: UInt64, body: String) throws {
        guard var list = data.branches[branch],
            let index = list.firstIndex(where: { $0.id == id })
        else { return }
        list[index].body = body
        data.branches[branch] = list
        try save()
    }

    /// Digest of every draft on this branch — the "Copy all" affordance.
    public func digest() -> String {
        formatDigest(
            comments.map {
                CommentContext(
                    path: $0.path, line: $0.line, lineText: $0.lineText,
                    body: $0.body, author: nil)
            })
    }

    private func save() throws {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        try encoder.encode(data).write(to: file)
    }
}

extension String {
    fileprivate func trimmingTrailingWhitespace() -> String {
        var view = self[...]
        while let last = view.last, last.isWhitespace {
            view = view.dropLast()
        }
        return String(view)
    }
}

File: core/src/comments.rs
//! Review comments: local drafts (notes-to-agent, never posted) and the
//! clipboard formats that feed them to a coding agent.
//!
//! Local comments live under `.git/preceipts/comments.json`, keyed by
//! branch — they are review state, not repo content, and they survive the
//! app closing but not intentional cleanup. GitHub feedback (fetched
//! separately) renders through the same `CommentContext` so every comment
//! in the app has the same "Copy to Clipboard" affordance.

use std::collections::BTreeMap;
use std::path::{Path, PathBuf};
use std::time::{SystemTime, UNIX_EPOCH};

use serde::{Deserialize, Serialize};

use crate::error::CoreError;

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum Side {
    Old,
    New,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct LocalComment {
    pub id: u64,
    pub path: String,
    /// 1-based line number on `side` of the diff.
    pub line: u32,
    pub side: Side,
    /// The diff line's text at the time the comment was written — quoted in
    /// the clipboard format and used to flag the comment as outdated later.
    pub line_text: String,
    pub body: String,
    /// Unix seconds; display formatting is the UI's job.
    pub created_at: u64,
}

/// Everything needed to render one comment to the clipboard, regardless of
/// origin (local draft or GitHub).
pub struct CommentContext<'a> {
    pub path: &'a str,
    pub line: u32,
    pub line_text: &'a str,
    pub body: &'a str,
    /// None for local drafts; "octocat" / "coderabbit[bot]" for GitHub.
    pub author: Option<&'a str>,
}

/// One comment as a markdown block with file:line context:
///
/// ```text
/// apps/next/components/foo.tsx:123
/// > const x = useMemo(...)
/// This memo is unnecessary — props are primitives.
/// ```
pub fn format_comment(comment: &CommentContext) -> String {
    let mut out = format!("{}:{}\n", comment.path, comment.line);
    if !comment.line_text.trim().is_empty() {
        out.push_str(&format!("> {}\n", comment.line_text.trim_end()));
    }
    if let Some(author) = comment.author {
        out.push_str(&format!("— {author}: "));
    }
    out.push_str(comment.body.trim());
    out.push('\n');
    out
}

/// Many comments as one digest, grouped by file — a paste-ready worklist
/// for a coding agent.
pub fn format_digest<'a>(comments: impl IntoIterator<Item = CommentContext<'a>>) -> String {
    let mut by_file: BTreeMap<String, Vec<String>> = BTreeMap::new();
    for comment in comments {
        by_file
            .entry(comment.path.to_string())
            .or_default()
            .push(format_comment(&comment));
    }
    let mut out = String::new();
    for (path, blocks) in by_file {
        out.push_str(&format!("## {path}\n\n"));
        for block in blocks {
            out.push_str(&block);
            out.push('\n');
        }
    }
    out.trim_end().to_string() + "\n"
}

#[derive(Default, Serialize, Deserialize)]
struct StoreFile {
    /// branch name → comments. Drafts are review state for a branch's PR.
    branches: BTreeMap<String, Vec<LocalComment>>,
    next_id: u64,
}

pub struct CommentStore {
    file: PathBuf,
    branch: String,
    data: StoreFile,
}

impl CommentStore {
    /// Open (or create) the store for a repo's git dir and current branch.
    pub fn open(git_dir: &Path, branch: &str) -> Result<CommentStore, CoreError> {
        let dir = git_dir.join("preceipts");
        std::fs::create_dir_all(&dir)?;
        let file = dir.join("comments.json");
        let data = match std::fs::read(&file) {
            Ok(bytes) => serde_json::from_slice(&bytes).unwrap_or_default(),
            Err(_) => StoreFile::default(),
        };
        Ok(CommentStore {
            file,
            branch: branch.to_string(),
            data,
        })
    }

    pub fn comments(&self) -> &[LocalComment] {
        self.data
            .branches
            .get(&self.branch)
            .map(Vec::as_slice)
            .unwrap_or(&[])
    }

    pub fn add(
        &mut self,
        path: &str,
        line: u32,
        side: Side,
        line_text: &str,
        body: &str,
    ) -> Result<&LocalComment, CoreError> {
        self.data.next_id += 1;
        let comment = LocalComment {
            id: self.data.next_id,
            path: path.to_string(),
            line,
            side,
            line_text: line_text.to_string(),
            body: body.to_string(),
            created_at: SystemTime::now()
                .duration_since(UNIX_EPOCH)
                .map(|d| d.as_secs())
                .unwrap_or(0),
        };
        let list = self.data.branches.entry(self.branch.clone()).or_default();
        list.push(comment);
        self.save()?;
        Ok(self.comments().last().expect("just pushed"))
    }

    pub fn remove(&mut self, id: u64) -> Result<(), CoreError> {
        if let Some(list) = self.data.branches.get_mut(&self.branch) {
            list.retain(|c| c.id != id);
        }
        self.save()
    }

    pub fn update_body(&mut self, id: u64, body: &str) -> Result<(), CoreError> {
        if let Some(list) = self.data.branches.get_mut(&self.branch) {
            if let Some(comment) = list.iter_mut().find(|c| c.id == id) {
                comment.body = body.to_string();
            }
        }
        self.save()
    }

    /// Digest of every draft on this branch — the "Copy all" affordance.
    pub fn digest(&self) -> String {
        format_digest(self.comments().iter().map(|c| CommentContext {
            path: &c.path,
            line: c.line,
            line_text: &c.line_text,
            body: &c.body,
            author: None,
        }))
    }

    fn save(&self) -> Result<(), CoreError> {
        let bytes = serde_json::to_vec_pretty(&self.data)?;
        std::fs::write(&self.file, bytes)?;
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn format_comment_includes_context() {
        let text = format_comment(&CommentContext {
            path: "src/foo.rs",
            line: 42,
            line_text: "let x = 1;  ",
            body: "  rename to count  ",
            author: None,
        });
        assert_eq!(text, "src/foo.rs:42\n> let x = 1;\nrename to count\n");
    }

    #[test]
    fn format_comment_attributes_github_authors() {
        let text = format_comment(&CommentContext {
            path: "a.ts",
            line: 7,
            line_text: "",
            body: "nit: prefer const",
            author: Some("coderabbit[bot]"),
        });
        assert_eq!(text, "a.ts:7\n— coderabbit[bot]: nit: prefer const\n");
    }

    #[test]
    fn digest_groups_by_file() {
        let digest = format_digest(vec![
            CommentContext {
                path: "b.rs",
                line: 2,
                line_text: "bar()",
                body: "extract helper",
                author: None,
            },
            CommentContext {
                path: "a.rs",
                line: 1,
                line_text: "foo()",
                body: "typo",
                author: None,
            },
            CommentContext {
                path: "a.rs",
                line: 9,
                line_text: "baz()",
                body: "dead code",
                author: None,
            },
        ]);
        assert!(digest.starts_with("## a.rs\n"));
        assert!(digest.contains("## b.rs\n"));
        let a_pos = digest.find("a.rs:1").unwrap();
        let a9_pos = digest.find("a.rs:9").unwrap();
        let b_pos = digest.find("b.rs:2").unwrap();
        assert!(a_pos < a9_pos && a9_pos < b_pos);
    }

    #[test]
    fn store_roundtrips_per_branch() {
        let dir = tempfile::tempdir().unwrap();
        let mut store = CommentStore::open(dir.path(), "feature-x").unwrap();
        store
            .add("src/a.rs", 10, Side::New, "let y = 2;", "why 2?")
            .unwrap();
        store
            .add("src/b.rs", 5, Side::Old, "old()", "this was removed — ok?")
            .unwrap();
        assert_eq!(store.comments().len(), 2);

        // Reopen: same branch sees both, another branch sees none.
        let store2 = CommentStore::open(dir.path(), "feature-x").unwrap();
        assert_eq!(store2.comments().len(), 2);
        let other = CommentStore::open(dir.path(), "main").unwrap();
        assert_eq!(other.comments().len(), 0);

        let digest = store2.digest();
        assert!(digest.contains("src/a.rs:10"));
        assert!(digest.contains("> let y = 2;"));
    }

    #[test]
    fn remove_and_update() {
        let dir = tempfile::tempdir().unwrap();
        let mut store = CommentStore::open(dir.path(), "b").unwrap();
        let id = store
            .add("f.rs", 1, Side::New, "x", "first")
            .unwrap()
            .id;
        store.update_body(id, "revised").unwrap();
        assert_eq!(store.comments()[0].body, "revised");
        store.remove(id).unwrap();
        assert!(store.comments().is_empty());
    }
}
